native/src/widget/scrollable.rs

```diff
@@ -736,227 +901,403 @@ pub fn draw<Renderer>(
-
-    if content_bounds.height > bounds.height {
-        let outer_width =
-            scrollbar_width.max(scroller_width) + 2 * scrollbar_margin;
-
-        let outer_bounds = Rectangle {
-            x: bounds.x + bounds.width - outer_width as f32,
-            y: bounds.y,
-            width: outer_width as f32,
-            height: bounds.height,
-        };
-
-        let scrollbar_bounds = Rectangle {
-            x: bounds.x + bounds.width
-                - f32::from(outer_width / 2 + scrollbar_width / 2),
-            y: bounds.y,
-            width: scrollbar_width as f32,
-            height: bounds.height,
+    shell: &mut Shell<'_, Message>,
+) {
+    if let Some(on_scroll) = on_scroll {
+        let delta_x = if content_bounds.width <= bounds.width {
+            0.0
+        } else {
+            state.scrollbar_x.map_or(0.0, |scrollbar| {
+                scrollbar.offset.absolute(
+                    Direction::Horizontal,
+                    bounds,
+                    content_bounds,
+                ) / (content_bounds.width - bounds.width)
+            })
         };
 
-        let ratio = bounds.height / content_bounds.height;
-        let scroller_height = bounds.height * ratio;
-        let y_offset = offset as f32 * ratio;
-
-        let scroller_bounds = Rectangle {
-            x: bounds.x + bounds.width
-                - f32::from(outer_width / 2 + scroller_width / 2),
-            y: scrollbar_bounds.y + y_offset,
-            width: scroller_width as f32,
-            height: scroller_height,
+        let delta_y = if content_bounds.height <= bounds.height {
+            0.0
+        } else {
+            state.scrollbar_y.map_or(0.0, |scrollbar| {
+                scrollbar.offset.absolute(
+                    Direction::Vertical,
+                    bounds,
+                    content_bounds,
+                ) / (content_bounds.height - bounds.height)
+            })
         };
 
-        Some(Scrollbar {
-            outer_bounds,
-            bounds: scrollbar_bounds,
-            scroller: Scroller {
-                bounds: scroller_bounds,
-            },
-        })
-    } else {
-        None
+        shell.publish(on_scroll(Vector::new(delta_x, delta_y)))
     }
 }
 
-fn notify_on_scroll<Message>(
-    state: &State,
-    on_scroll: &Option<Box<dyn Fn(f32) -> Message + '_>>,
-    bounds: Rectangle,
-    content_bounds: Rectangle,
-    shell: &mut Shell<'_, Message>,
-) {
-    if content_bounds.height <= bounds.height {
-        return;
-    }
-
-    if let Some(on_scroll) = on_scroll {
-        shell.publish(on_scroll(
-            state.offset.absolute(bounds, content_bounds)
-                / (content_bounds.height - bounds.height),
-        ));
-    }
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+/// The direction of the [`Scrollable`].
+pub enum Direction {
+    /// X or horizontal
+    Horizontal,
+    /// Y or vertical
+    Vertical,
 }
```

Comment:
I don't think this type is necessary. We always statically know the `Direction` of the scrollbar we are dealing with in the code.